### ISSUE TITLE
Instrumentation

### ISSUE THREAD

**Issue by ghaerr** [tkchia/gcc-ia16#111]
Hello @tkchia,

After reading a most interesting [article on logging C functions](http://www.justine.lol/ftrace/), written by the author of the [Cosmopolitan](https://github.com/jart/cosmopolitan) project, which does amazing things with the GCC compiler, I have been thinking that I'd like to add a similar logging capability to applications written for the ELKS system. I am interested to know what facilities might be already in place, or possibly added to, ia16-elf-gcc. I have come to believe that compiler function instrumentation could greatly help understand what's happening under the hood with many C programs.

The Cosmopolitan project uses gcc's `-pg`, `-mfentry` and `-mnop-mcount` options to insert a set of NOPs at the start of each profiled function, before any code is produced. When function tracing is desired, some C runtime code searches for the NOP signature and inserts a call to a function trace routine, which uses the stack return chain and ELF symbol data in order to work. 

For ELKS with ia16-elf-gcc, I've only been able to find the `-finstrument-functions` option, which generates code to push the address of the function being called and its call site. I've got that working, using `ia16-elf-nm` data from the ELF binary prior to being converted with `elf2elks`, to display function tracing.

For example, here's the function trace of the ELKS BASIC interpreter running on ELKS, tracing the statement "a = 3", and then exiting. The stack usage per function call and max stack overall is also calculated:
```
# basic
>main, from _start+19, stack 0/0
|>tty_isig, from main+22, stack 28/28
|>reset, from main+30, stack 10/28
ELKS BASIC
|>host_outputFreeMem, from main+42, stack 11/28
10240 bytes free

|>loop, from main+84, stack 136/136
||>host_readLine, from loop+5a, stack 141/141
|||>tty_isig, from host_readLine+13, stack 167/167
|||>fread, from fgetc+5f, stack 165/167
a = 3
|||>tty_raw, from host_readLine+66, stack 165/167
||>tokenize, from loop+a5, stack 143/167
|||>nextToken, from tokenize+27, stack 173/173
|||>nextToken, from tokenize+27, stack 173/173
|||>nextToken, from tokenize+27, stack 173/173
|||>nextToken, from tokenize+27, stack 173/173
||>tty_raw, from loop+b8, stack 160/173
||>processInput, from loop+bf, stack 144/173
|||>getNextToken, from processInput+1b, stack 152/173
|||>parseStmts, from processInput+65, stack 149/173
||||>parseAssignment, from parseStmts+71, stack 165/173
|||||>getNextToken, from parseAssignment+85, stack 171/173
|||||>getNextToken, from parseAssignment+164, stack 171/173
|||||>parseExpression, from parseAssignment+167, stack 169/173
||||||>parsePrimary, from parseExpression+12, stack 175/175
|||||||>parseNumberExpr, from parsePrimary+c8, stack 179/179
||||||||>getNextToken, from parseNumberExpr+1c, stack 185/185
||||||>parseBinOpRHS, from parseExpression+22, stack 182/185
|||||||>getTokPrecedence, from parseBinOpRHS+1b, stack 186/186
|||>getNextToken, from processInput+b4, stack 150/186
|||>parseStmts, from processInput+d7, stack 149/186
||||>parseAssignment, from parseStmts+71, stack 165/186
|||||>getNextToken, from parseAssignment+85, stack 171/186
|||||>getNextToken, from parseAssignment+164, stack 171/186
|||||>parseExpression, from parseAssignment+167, stack 169/186
||||||>parsePrimary, from parseExpression+12, stack 175/186
|||||||>parseNumberExpr, from parsePrimary+c8, stack 179/186
||||||||>stackPushNum, from parseNumberExpr+3b, stack 186/186
||||||||>getNextToken, from parseNumberExpr+1c, stack 185/186
||||||>parseBinOpRHS, from parseExpression+22, stack 182/186
|||||||>getTokPrecedence, from parseBinOpRHS+1b, stack 186/186
|||||>stackPopNum, from parseAssignment+183, stack 171/186
|||||>storeNumVariable, from parseAssignment+18c, stack 177/186
||||||>findVariable, from storeNumVariable+34, stack 186/186
Ok

|>loop, from main+84, stack 136/186
||>host_readLine, from loop+5a, stack 141/186
|||>tty_isig, from host_readLine+13, stack 167/186
|||>fread, from fgetc+5f, stack 165/186
|||>tty_raw, from host_readLine+4d, stack 165/186
|>tty_normal, from main+8f, stack 5/186
# 
```
The next step might be to add a mechanism that allows the entire C library and any application to be automatically compiled with function tracing, which could be turned on or off at runtime for debugging or inspection.

The problem is, `-finstrument-functions` generates quite a bit of code, for pushing the called function as well as its call site. It doesn't need to, as this could be calculated using the BP stack chain. It also produces both entry and exit calls, which combined with the speed and size issues, limits its use greatly. It would be really nice to have a way to generate NOPs at function start instead.

I notice that gcc has a `-fpatchable-function-entry=N,M`, which seemingly could do the job. However, this option is missing in ia16-elf-gcc.

What do you think? Are there other mechanisms that could work to help implement this capability? Could `-fpatchatble-function-entry=3` be easily implemented?

(On another note, this feature also requires the ability to either not run `elf2elks`, or keep a copy of the ld.gold's ELF output while otherwise completing the link. This was discussed in https://github.com/tkchia/gcc-ia16/issues/76#issuecomment-837074403 and I plan on re-opening that issue).

Thank you!

**Comment by ghaerr** [comment]
Hello @tkchia,

An update on ia16-elf-gcc instrumentation: 

Not being familiar enough with gcc internals, it seems most of the x86 instrumentation functions I was writing about above  are implemented in gcc-src/gcc/config/i386/i386.c, which I had thought was included in ia16-elf-gcc. I now realize only the gcc-src/gcc/config/ia16/ code is included, which does not include instrumentation. A quick review of the possibility of adding -mfentry or -mnop-count to ia16-elf-gcc looks like quite a bit of work.

The `-finstrument-functions` implementation is in the core compiler code, gcc-src/gcc/gimplify.c::gimplify_function_tree(), and is implemented using calls to two builtin functions BUILTIN_FUNCTION_PROFILE_ENTER/EXIT. In the current implementation, these functions are called near the function entry and exit (but not before any prologue code), and code is generated to push two arguments, which significantly increase the code size and slows execution. The arguments are the address of the function being called, and the call site address. Generating that additional code significantly disturbs the code generated for the rest of the function.

I have modified the compiler implementation to generate just a (three-byte) function call with no parameters for function entry and exit, and have the called function profiler dynamically determine the calling function and call site, using the return address and BP back chain information, combined with inspecting the caller function prologue code to see if BP, SI or DI are saved, to find the call site return address on the stack.

Would you be interested in seeing this modification? I am thinking that, by adding another command line option, the differentiation between the instrumentation options could be selected (for compatibility reasons) and this change might be able to be included in the ia16-elf-gcc. If included, it would enable ELKS programs to be efficiently instrumented, which then allows more understanding of program execution and timing.

Thank you!

**Comment by asiekierka** [comment]
I'd be interested in seeing it myself - on bare metal (or DOS?), I could allocate an interrupt for profiling and collect information this way...

**Comment by ghaerr** [comment]
Ok, @asiekierka, I will post what I have after cleaning things up a bit. I'm still working on perfecting some of the stack dump routines.

> on bare metal (or DOS?), I could allocate an interrupt for profiling and collect information this way...

That kind of profiling doesn't need these compiler generated calls - each interrupt would just record CS:IP and save the data for later matching with a symbol map file. The instrumentation I'm working is more suited to executing code at runtime as the result of the generated profile calls from each function - for instance a stack backtrace. For instance, here's the currently working stack dump code, which should work fine on bare metal, providing a symbol map file is available:
```
ELKS 0.6.0

login: root
# basic
AX=0000 BX=7460 CX=0001 DX=7464 SI=7460 DI=3448 BP=7446 SP=7422
CS:IP=5540:36f0 DS=6200 ES=6200 SS:SP=6200:7422

Level Addr    BP   DI   SI   Ret  Arg  Arg2 Arg3 Arg4
~~~~~ ~~~~    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   0: 7418 => 7446 3448 7460 36f7 dead 0000 0000 0000 0000 0000 0000 0000 0000 0000 print_stack (364e,83)
   1: 7446 => 744e 2c60 7460 3448 745c 2c60 2cb0 0019 0001 7460 7464 0000 0001 7472 __cyg_profile_func_enter+35 (36f7,83)
   2: 744e => 745c 2c60 2cb0 0019 0001 7460 7464 0000 0001 7472 0000 7478 7482 748d main+b (3448,83)
   3: 745c => 0000 0001 7472 0000 7478 7482 748d 7497 74a1 74b2 0000 6162 6973 0063 _start+19 (0019,81)
   4: 0000 => 0000 0000 5825 0900 2800 2900 2b00 2d00 2a00 2f00 3d00 3e00 3c00 3c00 _start+1 (0001,81)
>main, from _start+19, stack 0/0
AX=0019 BX=0000 CX=5401 DX=0000 SI=7460 DI=2ce5 BP=7414 SP=73f0
CS:IP=5540:36f0 DS=6200 ES=6200 SS:SP=6200:73f0

Level Addr    BP   DI   SI   Ret  Arg  Arg2 Arg3 Arg4
~~~~~ ~~~~    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   0: 73e6 => 7414 2ce5 7460 36f7 dead 00a1 0019 73fe 73fe 354e 7408 1742 1c9c 5014 print_stack (364e,83)
   1: 7414 => 7442 0000 7460 2ce5 0000 3448 0000 0000 0000 735f 6174 7472 312b 0000 __cyg_profile_func_enter+35 (36f7,83)
   2: 7442 => 744e 0000 7460 345b 0000 46b8 745c 2c60 2cb0 0019 0001 7460 7464 0000 tty_isig+b (2ce5,83)
   3: 744e => 745c 2c60 2cb0 0019 0001 7460 7464 0000 0001 7472 0000 7478 7482 748d main+1e (345b,83)
   4: 745c => 0000 0001 7472 0000 7478 7482 748d 7497 74a1 74b2 0000 6162 6973 0063 _start+19 (0019,81)
   5: 0000 => 0000 0000 5825 0900 2800 2900 2b00 2d00 2a00 2f00 3d00 3e00 3c00 3c00 _start+1 (0001,81)
|>tty_isig, from main+1e, stack 25/25
```

**Comment by asiekierka** [comment]
>That kind of profiling doesn't need these compiler generated calls

You still need to actually emit the interrupt on function entry/exit, though, no?

**Comment by ghaerr** [comment]
> You still need to actually emit the interrupt on function entry/exit, though?

That's kind of a neat idea, emitting an INT instruction at function prologue/epilogue. That would work and could use less space, at the cost of having to have an external profiler running on the hardware, unless the application or kernel also caught the interrupt at startup. The `-finstrument-functions` option, which already exists but is now modified, emits a call to a linked-in procedure from each compiled C function. This uses three bytes (and no arguments), so it's larger, and the profiler code is linked into the application.

**Comment by ghaerr** [comment]
Hello @asiekierka,

> I'd be interested in seeing it myself - on bare metal (or DOS?), I could allocate an interrupt for profiling and collect information this way...

I've posted the stack trace and instrumentation code (for running on ELKS), in https://github.com/jbruchon/elks/pull/1397. This code will produce the display first posted for this issue, when PROFILING is defined and the application is compiled with `-finstrument-functions`. That option causes ia16-elf-gcc to emit calls to `__cygwin_profile_func_enter` and `__cygwin_profile_func_exit` on function prologue and exit. However, the "stock" version of `-finstrument-functions` also emits code to pass two arguments to these functions, which aren't needed with my stack tracer, and increase the code emitted and slow down the runtime considerably. 

I have a compiler patch which I will publish shortly that causes ia16-elf-gcc to just emit the call instruction to the profiler functions, but without any passed arguments. I have a further modification in a separate `fixobj` program that effectively rewrites the generated .o file calls to the profiler with NOPs, so that there is very little runtime overhead, until the runtime re-inserts the calls to the profiler when profiling is requested. This is then essentially the same as later versions of gcc using the `-mfentry` and `-mnop-mcount` options. The `fixobj` program could also replace the profiler calls with `int $3` for your case of operating on bare metal, and not require any linked profiler code with the application.

Thank you!

**Comment by asiekierka** [comment]
Runtime re-insertion wouldn't be terribly helpful here, as the platform is ROM-based. However, thank you for the information!

**Comment by ghaerr** [comment]
Hello @tkchia,

Here is the patch required to change the functionality of `-finstrument-functions` to not generate code to push the two parameters normally required by the instrumentation functions that are normally declared:
```
void __cyg_profile_func_enter(void *arg1, void *arg2);
void  __cyg_profile_func_exit(void *arg1, void *arg2);
```
to 
```
void __cyg_profile_func_enter(void);
void  __cyg_profile_func_exit(void);
```
For ELKS, the code generated just after the function prologue and before the exit greatly interferes with the code generated for the function itself because of the parameter access, thus changing the nature of the profiling.

Patch for gcc-src/gcc/gimplify.c:
```
11718d11717
<       int no_args_to_instrument_fn = 1;
11729,11730c11728
<       if (!no_args_to_instrment_fn)
<         gimplify_seq_add_stmt (&cleanup, call);
---
>       gimplify_seq_add_stmt (&cleanup, call);
11733,11737d11730
<       if (no_args_to_instrument_fn) {
<         x = builtin_decl_implicit (BUILT_IN_PROFILE_FUNC_ENTER);
<         call = gimple_build_call (x, 0);
<         gimplify_seq_add_stmt (&body, call);
<       } else {
11748d11740
<       }
```
Here is gimplify.c with the patch applied: [gimplify.c.zip](https://github.com/tkchia/gcc-ia16/files/10296205/gimplify.c.zip)

Ideally, it would be better to have an option like `-finstrument-functions=2` or perhaps `-fia16-instrument-functions` that called a separately-defined BUILTIN_FUNC_ENTER/EXIT, perhaps named something like `__ia16_profile_func_enter` etc, if backwards compatibility with `-finstrument-functions` is desired. (The option generates so much code as to be nearly unusable, and I have found very little, if any modern systems that use this very old option).

If this modification were accepted, it would greatly help the ability to perform stack backtraces (with symbols) such as is currently coded and working with a patched compiler in https://github.com/jbruchon/elks/blob/master/elkscmd/debug/stacktrace.c.

Thank you!

**Comment by ghaerr** [comment]
Hello @tkchia,

I've been thinking about adding function tracing to the ELKS kernel. The problem is, that when using `-finstrument-functions-simple` and a far function, the compiler tries to emit a call to the near function `__cyg_profile_func_enter_simple` and fails:
```
#define INITPROC __far __attribute__ ((far_section, noinline, section (".fartext.init")))

void INITPROC foo(void)
{
}
```
```
far.c: In function ‘foo’:
far.c:5:1: error: cannot create relocatable call to far -> near thunk
 }
```
I was wondering how much work might be required to get this to work: instead of trying to get the compiler to emit a call to `__ia16_far_call_thunk.bx.0`, perhaps an additional builtin 'far' function could be created internally, (perhaps called `__cyg_profile_func_enter_simple_far`) with the result that it might be selected when the compiled function is far.

What do you think about that approach? I would suppose that `gimplify.c` would have to have knowledge about the current function's code segment attribute, and then emit code to the far version of the instrumentation. I could play around with this but wanted to see what you thought of the approach first, or whether there might be a simpler way.

Thank you!

**Comment by tkchia** [comment]
Hello @ghaerr,

You should be able to do something like this:
```
extern void __far __cyg_profile_func_enter_simple_far (void)
                  __attribute__ ((near_section));
extern void __far __cyg_profile_func_exit_simple_far (void)
                  __attribute__ ((near_section));

__attribute__ ((always_inline, gnu_inline, no_instrument_function))
extern inline void
__cyg_profile_func_enter_simple (void)
{
  __cyg_profile_func_enter_simple_far ();
}

__attribute__ ((always_inline, gnu_inline, no_instrument_function))
extern inline void
__cyg_profile_func_exit_simple (void)
{
  __cyg_profile_func_exit_simple_far ();
}
```
This will basically turn _all_ calls to `__cyg_profile_func_enter_simple ()` and `__cyg_profile_func_exit_simple ()`, into inlined calls to the far functions `__cyg_profile_func_enter_simple_far ()` and `__cyg_profile_func_exit_simple_far ()`.

The `near_section` attribute is not strictly needed, but may be good to use: if you call a `near_section` function from the default text segment, then the compiler will invoke it via a `push %cs` + near `call` (4 bytes), rather than a 5-byte far `call`.  Of course the callee actually needs to also be in the default text segment (e.g. itself defined as `near_section`) for this to work.

Thank you!

**Comment by ghaerr** [comment]
Hello @tkchia,

Thank you, excellent advice. I have played around with your suggestions and they work well. I hadn't known the power of using an inline function declaration to essentially redefine an internal (e.g. `__cyg_profile_func_enter_simple`) function to generate different code (or eliminate the calls altogether, if the inlined functions are empty). Very nice, and definitely beats considering a compiler modification! This is also useful to turn off the epilogue profiling call, which isn't needed.

> The near_section attribute is not strictly needed, but may be good to use

Testing this with some code in the ELKS kernel, specifically those declared with the `INITPROC` or `FATPROC` declarations (`__far __attribute__ ((far_section, noinline, section (".fartext.init")))`), the compiler ignores the `near_section` redefinition and produces an 5-byte `lcall` sequence anyways (as I suppose it has to). 

I'll do some more testing to determine whether to leave in the `near_section` attribute to save space - but am considering leaving it out so that all function prologues have a fixed 5-byte sequence (after the BP/SI/DI saves). I am considering whether to use a Cosmopolitan-style trick and rewriting the .o files or linked kernel to replace the lcalls with NOPs (equivalent to the the missing `-mfentry -mnop-mcount` in `ia16-elf-gcc`) to turn on and off profiling. I did port `fixobj.c` from Cosmopolitan as a test to rewrite `ia16-elf-gcc` .o files and got a test version of that working. That's probably overkill and a lot of work in order to get dynamic --ftrace-style function profiling in the kernel, rather than recompiling or having the function profiler just return when not profiling.

Thank you!
